service/csvService.go

```diff
@@ -6,133 +6,219 @@ import (
+func (c CsvServiceImpl) FilterByID(id string) ([]model.Doc, error) {
+	file, err := os.Open(getConfigService().GetConfig().CSV.FileName)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to open file: %v", err)
 	}
 	reader := csv.NewReader(file)
 	records, errReader := reader.ReadAll()
 	if errReader != nil {
 		return nil, errReader
 	}
 
-	var recordsStruct []Doc
+	var recordsStruct []model.Doc
+	conditions := make(map[string] string)
+	conditions["Key"] = id
 
 	for _, record := range records {
-		fmt.Println(record)
 
-		doc := Doc{
-			Key:       record[0],
-			Title:     record[1],
-			Type:      record[2],
-			Published: record[3],
-		}
-		if doc.Key == id {
-			recordsStruct = append(recordsStruct, doc)
+		// if record[0] != id {
+			// continue
+		// }
+		docRecord := toDoc(record)
+		fields := structs.Fields(docRecord)
+
+		if ( filterRecord(fields, conditions) ) {
+			recordsStruct = append(recordsStruct, toDoc(record))
 		}
+
 	}
 
 	if len(records) == 0 {
-		fmt.Println("The file is empty")
+		log.Info("The file is empty")
 		return nil, errors.New("The file is empty")
 	}
 	defer file.Close()
 
 	return recordsStruct, nil
 }
 
-// func fileExists(file string) bool {
-// 	if _, err := os.Stat(file); os.IsNotExist(err) {
-// 		fmt.Println("File "+file+" does not exists")
-// 		return false
-// 	}
-// 	return true
-// }
-
-// func readFileByRow(file string) ([]string, error) {
+func toDoc(record []string) model.Doc {
+	return model.Doc{
+		Key:       record[0],
+		Title:     record[1],
+		Type:      record[2],
+		Published: record[3],
+	}
+}
 
-// }
-func getAndSaveData() {
+func filterRecord(s []*structs.Field, m map[string]string) bool {
+	for _, f := range s {
+		log.Info("validating "+f.Name()+" in conditions "+m[f.Name()])
+		if filter := m[f.Name()]; filter != "" {
+			if f.Value().(string) != filter {
+				return false
+			}
+		}
 
-	response, err := http.Get("http://openlibrary.org/search.json?q=the+lord+of+the+rings&page=1")
+	}
+	return true
+}
 
+func deleteCsv() {
+	err := os.Remove(getConfigService().GetConfig().CSV.FileName)
 	if err != nil {
-		fmt.Println(err.Error())
-		panic(err)
+		log.Error("Cannot delete file", err)
 	}
+}
 
-	responseData, err := ioutil.ReadAll(response.Body)
+func getAndSaveData(topic string) error {
+	// the+lord+of+the+rings&page=1
+	// <search-term>  <page>
+
+	deleteCsv() // before write results clears file
+	escapedTopic := strings.ReplaceAll(topic, 	" ", "+")
+	apiUrl := strings.ReplaceAll(getConfigService().GetConfig().External.ApiUrl, "<topic>", escapedTopic)
+	file, err := openOrCreate(getConfigService().GetConfig().CSV.FileName)
 
 	if err != nil {
-		log.Fatal(err)
+		log.Error("Unable to use file", err)
+		return err
 	}
 
-	var responseObject SearchResponse
+	for i := 1; true; i++ {
+		response, err := http.Get(strings.ReplaceAll(apiUrl, "<page>", strconv.Itoa(i)))
+
+		if err != nil {
+			log.Info(err.Error())
```

Comment:
This is a log.Error(), right? as you're showing error information.

Reply:
Yes, fixing it.